### PR TITLE
Part 1 of 4: Auditing our required parameters for the C# SDK. 

### LINKED ISSUE #25: Allow users to omit parameters for optional defaults
# Customer Issue

Many Smartsheet API calls allow for customisation of query options through a wide variety of optional properties. This can be shown by looking at one of the SDK interfaces for accessing Smartsheet functionality: retrieving a sheet.

From the `SheetResources` interface in [smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs)

```
        Sheet GetSheet(
                    long sheetId,
                    IEnumerable<SheetLevelInclusion> includes,
                    IEnumerable<SheetLevelExclusion> excludes,
                    IEnumerable<long> rowIds,
                    IEnumerable<int> rowNumbers,
                    IEnumerable<long> columnIds,
                    long? pageSize,
                    long? page); 
```

The documentation for the parameters from `includes` onwards, denotes that the properties are optional as part of the task of retrieving a sheet e.g. `/// <param name="includes"> used to specify the optional objects to include. </param>`

The parameters are optional for the underlying http API call, but not for users of the C# SDK, who have to provide an input for all parameters, bar the final two, `pageSize`, and `page`, which are denoted as optional owing to them being a nullable long i.e. `long?`.

This results in invocations to the `GetSheet` implementation like this from https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/sdk-csharp-sample-net60/Program.cs#L30:

```
    var sheet = smartsheet.SheetResources.GetSheet(sheetId, null, null, null, null, null, null, null);
```

# Proposal

A review of the interfaces throughout the API should take place to establish which of the parameters are optional, and which are mandatory. After identifying genuinely optional parameters, we can modify existing interfaces and implementations to mark parameters as optional using the [Nullable operator](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nullable-value-types) `?`.

We should then expect to see invocations to Smartsheet SDK functionality that omit parameters where they are the default:

```
    var sheet = smartsheet.SheetResources.GetSheet(sheetId);
```

# Next Steps

Interfaces should be reviewed to consider which methods contain parameters which are permittable by the Smartsheet API to be `null` values.

The full list of interfaces to review:

- [x] [AttachmentVersioningResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/AttachmentVersioningResources.cs#L29)
- [x] [CommentAttachmentResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/CommentAttachmentResources.cs#L30)
- [x] [ContactResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/ContactResources.cs#L30)
- [x] [DiscussionAttachmentResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/DiscussionAttachmentResources.cs#L30)
- [x] [DiscussionCommentResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/DiscussionCommentResources.cs#L28)
- [x] [EventResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/EventResources.cs#L23)
- [x] [FavoriteResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/FavoriteResources.cs#L32)
- [x] [FolderResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/FolderResources.cs#L30)
- [x] [FolderSheetResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/FolderSheetResources.cs#L30)
- [x] [GroupResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/GroupResources.cs#L30)
- [x] [HomeFolderResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/HomeFolderResources.cs#L30)
- [x] [HomeResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/HomeResources.cs#L31)
- [x] [ImageUrlsResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/ImageUrlsResources.cs#L33)
- [x] [PassthroughResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/PassthroughResources.cs#L26)
- [x] [ReportResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/ReportResources.cs#L30)
- [x] [RowAttachmentResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/RowAttachmentResources.cs#L30)
- [x] [RowColumnResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/RowColumnResources.cs#L30)
- [x] [RowDiscussionResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/RowDiscussionResources.cs#L30)
- [x] [SearchResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SearchResources.cs#L31)
- [x] [ShareResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/ShareResources.cs#L32)
- [x] [SheetAttachmentResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetAttachmentResources.cs#L29)
- [x] [SheetAutomationRuleResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetAutomationRuleResources.cs#L32)
- [x] [SheetColumnResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetColumnResources.cs#L31)
- [x] [SheetCommentResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetCommentResources.cs#L29)
- [x] [SheetCrossSheetReferenceResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetCrossSheetReferenceResources.cs#L27)
- [x] [SheetDiscussionResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetDiscussionResources.cs#L29)
- [x] [SheetFilterResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetFilterResources.cs#L33)
- [x] [SheetResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetResources.cs#L33)
- [x] [SheetRowResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetRowResources.cs#L32)
- [x] [SheetSummaryResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetSummaryResources.cs#L30)
- [x] [SheetUpdateRequestResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SheetUpdateRequestResources.cs#L23)
- [x] [SightResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/SightResources.cs#L26)
- [x] [TemplateResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/TemplateResources.cs#L30)
- [x] [TokenResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/TokenResources.cs#L30)
- [x] [UserResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/UserResources.cs#L30)
- [x] [UserSheetResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/UserSheetResources.cs#L31)
- [x] [WebhookResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/WebhookResources.cs#L28)
- [x] [WorkspaceFolderResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/WorkspaceFolderResources.cs#L30)
- [x] [WorkspaceResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/WorkspaceResources.cs#L31)
- [x] [WorkspaceSheetResources](https://github.com/smartsheet/smartsheet-csharp-sdk/blob/mainline/smartsheet-csharp-sdk/main/Smartsheet/Api/WorkspaceSheetResources.cs#L30)

### PR DESCRIPTION
**Pull request by timothodge**
Looking to fix things up based on this open issue: https://github.com/smartsheet/smartsheet-csharp-sdk/issues/25

This is one quarter of the audit done. Need to continue to do the other 3/4 but wanted to make sure people liked this before I move on. Here is the public API docs for anyone who wants to cross check: https://smartsheet.redoc.ly/tag/groups/

### CODE REVIEW

File: smartsheet-csharp-sdk/main/Smartsheet/Api/AttachmentVersioningResources.cs [hunk 1]
```diff
@@ -47,7 +47,7 @@ public interface AttachmentVersioningResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        Attachment AttachNewVersion(long sheetId, long attachmentId, string file, string fileType);
+        Attachment AttachNewVersion(long sheetId, long attachmentId, string? file, string? fileType);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/AttachmentVersioningResources.cs]: Screenshot 2024-04-29 at 2 30 51 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/AttachmentVersioningResources.cs [hunk 2]
```diff
@@ -80,6 +80,6 @@ public interface AttachmentVersioningResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        PaginatedResult<Attachment> ListVersions(long sheetId, long attachmentId, PaginationParameters paging);
+        PaginatedResult<Attachment> ListVersions(long sheetId, long attachmentId, PaginationParameters? paging);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/AttachmentVersioningResources.cs]: Screenshot 2024-04-29 at 2 31 25 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/ContactResources.cs [hunk 3]
```diff
@@ -57,6 +57,6 @@ public interface ContactResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        PaginatedResult<Contact> ListContacts(PaginationParameters paging);
+        PaginatedResult<Contact> ListContacts(PaginationParameters? paging);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/ContactResources.cs]: Screenshot 2024-04-29 at 2 56 53 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/EventResources.cs [hunk 4]
```diff
@@ -44,7 +44,7 @@ public interface EventResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        EventResult ListEvents(object since, string streamPosition, int? maxCount, bool? numericDates);
+        EventResult ListEvents(object? since, string? streamPosition, int? maxCount, bool? numericDates);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/EventResources.cs]: Screenshot 2024-04-29 at 3 59 36 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/FavoriteResources.cs [hunk 5]
```diff
@@ -61,7 +61,7 @@ public interface FavoriteResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        PaginatedResult<Favorite> ListFavorites(PaginationParameters paging);
+        PaginatedResult<Favorite> ListFavorites(PaginationParameters? paging);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/FavoriteResources.cs]: Screenshot 2024-04-29 at 4 06 11 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/FolderResources.cs [hunk 6]
```diff
@@ -43,7 +43,7 @@ public interface FolderResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        Folder GetFolder(long folderId, IEnumerable<FolderInclusion> include);
+        Folder GetFolder(long folderId, IEnumerable<FolderInclusion>? include);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/FolderResources.cs]: Screenshot 2024-04-29 at 4 12 46 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/FolderResources.cs [hunk 7]
```diff
@@ -94,7 +94,7 @@ public interface FolderResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        PaginatedResult<Folder> ListFolders(long folderId, PaginationParameters paging);
+        PaginatedResult<Folder> ListFolders(long folderId, PaginationParameters? paging);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/FolderResources.cs]: Screenshot 2024-04-29 at 4 13 07 PM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/FolderResources.cs [hunk 8]
```diff
@@ -133,7 +133,7 @@ public interface FolderResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        Folder CopyFolder(long folderId, ContainerDestination destination, IEnumerable<FolderCopyInclusion> include, IEnumerable<FolderRemapExclusion> skipRemap);
+        Folder CopyFolder(long folderId, ContainerDestination destination, IEnumerable<FolderCopyInclusion>? include, IEnumerable<FolderRemapExclusion>? skipRemap);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/FolderResources.cs:136]: Screenshot 2024-04-29 at 4 13 38 PM

**Review by priyankalakhe — APPROVED** [review]

File: smartsheet-csharp-sdk/main/Smartsheet/Api/FolderSheetResources.cs [hunk 9]
```diff
@@ -59,7 +59,7 @@ public interface FolderSheetResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        Sheet CreateSheetFromTemplate(long folderId, Sheet sheet, IEnumerable<TemplateInclusion> includes);
+        Sheet CreateSheetFromTemplate(long folderId, Sheet sheet, IEnumerable<TemplateInclusion>? includes);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/FolderSheetResources.cs]: Screenshot 2024-04-30 at 8 43 21 AM

File: smartsheet-csharp-sdk/main/Smartsheet/Api/FolderSheetResources.cs [hunk 10]
```diff
@@ -77,7 +77,7 @@ public interface FolderSheetResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        Sheet ImportCsvSheet(long folderId, string file, string sheetName, int? headerRowIndex, int? primaryColumnIndex);
+        Sheet ImportCsvSheet(long folderId, string file, string? sheetName, int? headerRowIndex, int? primaryColumnIndex);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/FolderSheetResources.cs]: Screenshot 2024-04-30 at 8 43 30 AM

There is a need to update the public API docs. Issue listed here: [#61](https://github.com/smartsheet/smartsheet-csharp-sdk/issues/61)

File: smartsheet-csharp-sdk/main/Smartsheet/Api/GroupResources.cs [hunk 11]
```diff
@@ -42,7 +42,7 @@ public interface GroupResources
         /// <exception cref="ResourceNotFoundException"> if the resource cannot be found </exception>
         /// <exception cref="ServiceUnavailableException"> if the REST API service is not available (possibly due to rate limiting) </exception>
         /// <exception cref="SmartsheetException"> if there is any other error during the operation </exception>
-        PaginatedResult<Group> ListGroups(PaginationParameters paging);
+        PaginatedResult<Group> ListGroups(PaginationParameters? paging);
```
Comment (timothodge) [inline smartsheet-csharp-sdk/main/Smartsheet/Api/GroupResources.cs]: Screenshot 2024-04-30 at 10 47 41 AM

**Review by rzaracota — APPROVED** [review]

**Pull request merged**